tests/data/test_sampler.py

```diff
@@ -0,0 +1,150 @@
+    kwargs = {}
+    kwargs["n_batches"] = draw(st.integers(10, 20))
+    kwargs["batch_size"] = draw(st.integers(2, 16))
+    kwargs["full_last_batch"] = draw(st.booleans())
+
+    # indices = [0, ..., (n_batches*batch_size)-1-int(not(full_last_batch))]
+    indices = list(range(kwargs["n_batches"] * kwargs["batch_size"]))
+    if not kwargs["full_last_batch"]:
+        del indices[-1]
+
+    if not return_kwargs:
+        return indices
+    return indices, kwargs
+
+
+# Tests -----------------------------------------------------------------------
+
+
+@given(dataset_kwargs=dataset_gen(return_kwargs=True))
+def test_sorta_grad_correct_len(dataset_kwargs: Tuple[List, Dict]):
+    dataset, kwargs = dataset_kwargs
+
+    dataset = sorted(dataset)
+    sampler = SortaGrad(
+        dataset,
+        drop_last=kwargs["full_last_batch"],
+        batch_size=kwargs["batch_size"],
+        shuffle=False,
+    )
+    assert len(sampler) == kwargs["n_batches"]
+
+
+@given(dataset_kwargs=dataset_gen(return_kwargs=True))
+def test_sorta_grad_batches_non_empty(dataset_kwargs: Tuple[List, Dict]):
+    dataset, kwargs = dataset_kwargs
+
+    dataset = sorted(dataset)
+    sampler = SortaGrad(
+        dataset,
+        drop_last=kwargs["full_last_batch"],
+        batch_size=kwargs["batch_size"],
+        shuffle=False,
+    )
+    for batch in sampler:
+        assert len(batch) > 0
+
+
+@given(dataset_kwargs=dataset_gen(return_kwargs=True))
+def test_sorta_grad_first_pass_sequential_remaining_random(
+    dataset_kwargs: Tuple[List, Dict]
+):
+    dataset, kwargs = dataset_kwargs
+
+    dataset = sorted(dataset)
+    sortagrad = SortaGrad(
+        dataset,
+        drop_last=kwargs["full_last_batch"],
+        batch_size=kwargs["batch_size"],
+        shuffle=True,
+    )
+
+    indices: list
+    for pass_ in range(100):
+        indices = []
+        for batch in sortagrad:
+            indices.extend(batch)
+
+        assert sorted(indices) == dataset
+
+        if pass_ == 0:
+            assert indices == sorted(indices)
+        else:
+            assert indices != sorted(indices)
+
+
+@given(
+    dataset_kwargs=dataset_gen(return_kwargs=True),
+    n_iterators=st.integers(min_value=1, max_value=10),
+    n_sequential=st.integers(min_value=1, max_value=10),
+    max_sequential=st.integers(min_value=11, max_value=20),
+)
+def test_sequential_strategy_seq_iter_when_epoch_in_seq_epochs(
+    dataset_kwargs: Tuple[List, Dict],
+    n_iterators: int,
+    n_sequential: int,
+    max_sequential: int,
+):
+    dataset, kwargs = dataset_kwargs
+
+    dataset_batches = []
+    batch = []
+    for elem in dataset:
+        batch.append(elem)
+        if len(batch) == kwargs["batch_size"]:
+            dataset_batches.append(batch)
+            batch = []
+    if batch and not kwargs["full_last_batch"]:
+        dataset_batches.append(batch)
+    sequential = set(
+        sorted(random.sample(range(max_sequential), n_sequential))
+    )
```

Comment:
`max_sequential` should actually be a set generated by Hypothesis rather than an integer that is then used internally to generate a set? This way Hypothesis can control `n_sequential` and shrink both values down when failing to the minimal possible test case.

Reply:
I have added a "given" parameter to that test, called `sequential`, and it uses the Hypothesis `lists` function to generate a list of random numbers in a specific range. Let me know if this solution could be fine or if I have to change it

Reply:
I have created a separate function that returns a `SearchStrategy` for the list of sequential epoch numbers, so `max_sequential` and `n_sequential` are not needed anymore in the test parameters. Let me know if this solution could be fine or if I need to change it